src/actions/measurements.ts

```diff
@@ -611,14 +612,23 @@ export const applyMeasurementsColorBy = (
     }
   }
 
-  if (controls.measurementsColorGrouping !== undefined) {
-    dispatch({type: REMOVE_METADATA, nodeAttrsToRemove: [encodeMeasurementColorBy(controls.measurementsColorGrouping)]});
-  }
-  if (controls.measurementsColorGrouping !== groupingValue) {
-    dispatch({type: CHANGE_MEASUREMENTS_COLOR_GROUPING, controls:{measurementsColorGrouping: groupingValue}});
-  }
-  dispatch({type: ADD_EXTRA_METADATA, newNodeAttrs, newColorings});
-  dispatch(changeColorBy(measurementColorBy));
+  /**
+   * Batching all dispatch actions together to prevent multiple renders
+   * This is also _required_ to prevent error in calcColorScale during extra renders:
+   * 1. REMOVE_METADATA removes current measurements coloring from metadata.colorings
+   * 2. This triggers the componentDidUpdate in controls/color-by, which dispatches changeColorBy.
+   * 3. calcColorScale throws error because the current coloring is no longer valid as it was removed by REMOVE_METADATA in step 1.
+   */
+  batch(() => {
```

Comment:
I didn't know about `batch` - was a good opportunity to catch up with some of the many changes in the redux/react world. [It seems that React 18 will do this automatically](https://react-redux.js.org/api/batch) (react 18 upgrade PR: https://github.com/nextstrain/auspice/pull/1659). I wondered why you didn't just add the new node info, change colour, then delete the old data... however since the `node.attrs` key names are reused I can imagine this causes some undesirable side-effects?

Reply:
> I wondered why you didn't just add the new node info, change colour, then delete the old data... however since the node.attrs key names are reused I can imagine this causes some undesirable side-effects?

Yeah, things can get complicated if the node.attrs key names are the same but the underlying values changed. A concrete example is changing measurements collections. They might have the same reference strain (which means they have the same node.attrs key), but completely different measurements so the data needs to be completely updated. It's cleaner to remove old data then apply the new data instead of trying to keep track of which nodes needs to be updated.